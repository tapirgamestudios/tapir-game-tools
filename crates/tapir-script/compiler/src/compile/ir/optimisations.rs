use std::{
    collections::HashMap,
    ops::{BitOr, BitOrAssign},
};

use petgraph::visit::Dfs;

use crate::{ast::SymbolId, compile::ir::TapIrFunction};

mod empty_phi;

#[derive(Clone, Copy, PartialEq, Eq)]
enum OptimisationResult {
    DidSomething,
    DidNothing,
}

impl BitOrAssign for OptimisationResult {
    fn bitor_assign(&mut self, rhs: Self) {
        *self = *self | rhs;
    }
}

impl BitOr for OptimisationResult {
    type Output = OptimisationResult;

    fn bitor(self, rhs: Self) -> Self::Output {
        if self == OptimisationResult::DidSomething || rhs == OptimisationResult::DidSomething {
            OptimisationResult::DidSomething
        } else {
            OptimisationResult::DidNothing
        }
    }
}

fn rename_all_variables(
    function: &mut TapIrFunction,
    renames: &HashMap<SymbolId, SymbolId>,
) -> OptimisationResult {
    if renames.is_empty() {
        return OptimisationResult::DidNothing;
    }

    let mut did_something = OptimisationResult::DidNothing;

    let mut dfs = Dfs::new(&*function, function.root);
    while let Some(block_id) = dfs.next(&*function) {
        let block = function.block_mut(block_id).expect("Could not find block");

        for symbol in block.sources_mut() {
            if let Some(renamed_symbol) = renames.get(symbol) {
                *symbol = *renamed_symbol;
                did_something = OptimisationResult::DidSomething;
            }
        }

        for symbol in block.targets_mut() {
            if let Some(renamed_symbol) = renames.get(symbol) {
                *symbol = *renamed_symbol;
                did_something = OptimisationResult::DidSomething;
            }
        }
    }

    did_something
}
