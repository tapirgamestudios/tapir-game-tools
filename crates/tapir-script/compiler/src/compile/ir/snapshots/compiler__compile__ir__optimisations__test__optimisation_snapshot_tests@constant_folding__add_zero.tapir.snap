---
source: crates/tapir-script/compiler/src/compile/ir/optimisations.rs
expression: output
input_file: crates/tapir-script/compiler/src/compile/ir/snapshot_tests/optimisations/constant_folding/add_zero.tapir
---
----------- before -------------
=======================
fn @toplevel() -> 

---- block 0 ----
getprop temp.14, 0
temp.15 = int 0
temp.16 = temp.14 + temp.15
storeprop temp.16, 0
wait
temp.17 = int 0
getprop temp.18, 0
temp.19 = temp.17 + temp.18
storeprop temp.19, 0
wait
temp.20 = fix 0
getprop temp.21, 1
temp.22 = temp.20 + temp.21
storeprop temp.22, 1
wait
getprop temp.23, 1
temp.24 = fix 0
temp.25 = temp.23 + temp.24
storeprop temp.25, 1
return 


----------- optimised -------------
=======================
fn @toplevel() -> 

---- block 0 ----
getprop temp.14, 0
storeprop temp.14, 0
wait
getprop temp.18, 0
storeprop temp.18, 0
wait
getprop temp.21, 1
storeprop temp.21, 1
wait
getprop temp.23, 1
storeprop temp.23, 1
return
